packages/core/src/baseclient.ts

```diff
@@ -417,8 +417,8 @@ export abstract class BaseClient<B extends Backend, O extends Options> implement
     const options = this.getOptions();
     const { environment, release, dist, maxValueLength = 250 } = options;
 
-    if (!('environment' in event)) {
-      event.environment = 'environment' in options ? environment : 'production';
+    if (event.environment === undefined && environment !== undefined) {
+      event.environment = environment;
```

Comment:
Is this related to tracestate?

Reply:
Yes. Environment is part of the data that goes into tracestate, so it needs to be set in time for that to happen. If the user provides an environment, we're already good, but if we need to use the default, at the moment that's not getting added until we're sending the event. This moves it earlier so that it's available when we compute the `tracestate` value. 

See https://github.com/getsentry/sentry-javascript/pull/3242.